[RDY] Refactor event loop layer

tarruda opened the pull request:
Like #2650, this extracts some changes from my private #2371/fix branch. This PR separates the event loop layer from the rest of the code and builds a small abstraction around libuv/watcher API, which is done to keep some sanity for when the event loop runs in a separate thread. The goal is to contain multithreaded code under src/nvim/event and leave the rest as untouched as possible, with minimum amount of data/modules shared between threads.

oni-link commented on src/nvim/event/rstream.c:
```diff
@@ -5,148 +5,78 @@
+    // be processed between reads.
+    uv_idle_init(UV(loop), &rstream->uv.idle);
+    rstream->uv.idle.data = NULL;
+    rstream->uv.idle.data = rstream;
```
Previous line could be removed.

tarruda replied:
:+1:

tarruda committed:
test: Increase determinism in screen_basic_spec tests

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -262,7 +263,7 @@ static void read_cb(RStream *rstream, RBuffer *buf, void *data, bool eof)
+  } while (rbuffer_size(input->read_stream.buffer));
 
   // Make sure the next input escape sequence fits into the ring buffer
   // without wrap around, otherwise it could be misinterpreted.
```
`rbuffer_reset(input->read_buffer);` needs to be replaced with `rbuffer_reset(input->read_stream.buffer);` in the following line.

tarruda replied:
:+1:

oni-link commented on src/nvim/event/loop.c:
```diff
@@ -0,0 +1,43 @@
+
+void loop_run(Loop *loop)
+{
+  uv_run(&loop->uv, UV_RUN_DEFAULT);
```
Is the return value intentionally dropped?

tarruda replied:
Not intentionally, for now this is just a stub module that I plan to finish later

oni-link commented on src/nvim/os/job.c:
```diff
@@ -28,25 +27,21 @@
-        uv_close((uv_handle_t *)job->proc_std##stream, close_cb);          \
-      }                                                                    \
+    if (!job->stream.closed) {                                             \
+      stream_close(&job->stream, on_##stream_close);                       \
```
`on_##stream_close` -> `on_stream_close`

tarruda replied:
:+1:

oni-link commented on src/nvim/os/job.c:
```diff
@@ -117,63 +112,45 @@ Job *job_start(JobOptions opts, int *status)
+      uv_close((uv_handle_t *)job->proc_stderr, NULL);
     }
     process_close(job);
     event_poll(0);
```
`event_poll(0)` should be called before `process_close(job)`, to prevent freeing the job structure too soon.
Without this change the following patch (simulates an error while spawning a pty) and usage of `:term` should produce an crash:
```diff
diff --git a/src/nvim/os/pty_process.c b/src/nvim/os/pty_process.c
index cb59355..ebffac1 100644
--- a/src/nvim/os/pty_process.c
+++ b/src/nvim/os/pty_process.c
@@ -72,6 +72,7 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
     init_child(job);
     abort();
   }
+  goto error;
 
   // make sure the master file descriptor is non blocking
   int master_status_flags = fcntl(master, F_GETFL);
```

tarruda replied:
I'm currently refactoring the job module in a similar way that the server/rstream/wstream modules were refactored in this PR. I will take this into consideration for the refactored version, thanks

tarruda replied:
This should no longer be a problem, the process memory is now managed by the caller

oni-link commented on src/nvim/event/pty_process.c:
```diff
@@ -255,3 +202,50 @@ static void init_termios(struct termios *termios) FUNC_ATTR_NONNULL_ALL
+  return true;
+}
+
+static void chld_handler(uv_signal_t *handle, int signum)
```
Are UvProcesses on a different loop than PtyProcesses? Otherwise an UvProcess could be handled here and in its exit callback (also two calls to `internal_exit_cb()`).

tarruda replied:
They are handled in the same loop, but it's only possible to have `waitpid` return a pid > 0 once(at least in my tests). So if libuv `SIGCHLD` handler receives it first, this one will short circuit because the process will have already exited. The same will happen for libuv if we catch it first: https://github.com/libuv/libuv/blob/v1.x/src/unix/process.c#L78-L80

oni-link commented on src/nvim/event/pty_process.c:
```diff
@@ -156,31 +98,36 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
   return false;
```
After `kill(pid, SIGKILL)` we should call `waitpid(pid,NULL,0)`.

tarruda replied:
:+1:

oni-link commented on src/nvim/eval.c:
```diff
@@ -10681,29 +10691,31 @@ static void f_jobclose(typval_T *argvars, typval_T *rettv)
     if (!strcmp(stream, "stdin")) {
-      job_close_in(job);
-      ((TerminalJobData *)job_data(job))->stdin_closed = true;
+      stream_close(proc->in, NULL);
```
`jobclose()` could be called for the same stream multiple times. `stream_close()` would `assert()`, but `process_close_{in,out,err,streams}` should be safe to use here and in the following cases.

tarruda replied:
:+1:

oni-link commented on src/nvim/eval.c:
```diff
@@ -15214,18 +15216,14 @@ static void f_termopen(typval_T *argvars, typval_T *rettv)
     return;
   }
-  TerminalJobData *data = opts.data;
   TerminalOptions topts = TERMINAL_OPTIONS_INIT;
```
Initialization should not be necessary. All fields are set in the following lines.

tarruda replied:
:+1:

oni-link commented on src/nvim/event/pty_process.c:
```diff
@@ -20,92 +20,39 @@
+    init_child(ptyproc);
     abort();
   }
 
```
Adding here a `goto error;` leads to a seg fault when using `:term`, because later on the uninitialized callback pointer `internal_close_cb` is used in `pty_process_close()`.

tarruda replied:
:+1:

oni-link commented on src/nvim/eval.c:
```diff
@@ -10724,15 +10734,13 @@ static void f_jobsend(typval_T *argvars, typval_T *rettv)
   }
 
-  if (((TerminalJobData *)job_data(job))->stdin_closed) {
+  if (data->stdin_closed) {
```
Do we still need `stdin_closed` in `TerminalJobData`?
We could now do something like this:
```c
  Process *proc = (Process *)&data->proc;
  if (!proc->in || proc->in->closed) {
```

tarruda replied:
:+1:

tarruda committed:
test: lower sleep value in job test

Since sleep is a grandchild of nvim, it is not killed after the test ends.
Using a low sleep value allows it to exit automatically after a small interval.

tarruda commented:
Been trying to find out why the OSX/travis tests are hanging on this PR, and so far had little success. I cannot reproduce it running on a local OSX vm, and I don't think it is related to OSX.

Here's what I know so far:

- It happens with the job_wait/timeout test
- It is caused by the `sleep 10000` shell command which somehow causing the travis terminal/shell hang after the tests exit
- Using a lower sleep value, 10 for example, allows the build to continue after the sleep exits
- I can confirm that, in both linux and OSX, this PR and the master branch, the sleep process sticks around after the tests(until killed or the terminal is closed). In all cases nvim kills all children on exit(in this case, sleep is a grandchild)

Considering these facts, it seems the reasonable thing to do is to use a smaller sleep value, which would not only fix the "zombie" sleep processes sticking around after tests, but also make the OSX/travis build pass on this PR. 

What's messing with my head is trying to figure out why this doesn't happen on master, I simply cannot see how the process cleanup refactor would cause this issue on OSX/travis, especially because I already verified that in all branches/platforms the result is the same: The sleep process sticks after the tests.

I'd love to hear ideas about what is causing this, but in any case I'm not going to let this stop me from proceeding, I will just lower the sleep value and finish the PR.

justinmk commented:
I didn't see the OSX logs before you updated this PR, but we have been seeing random hangs on OSX on master. I just assumed it was a travis performance issue as it was in May.

oni-link commented on src/nvim/event/pty_process.c:
```diff
@@ -151,36 +93,44 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
-  job_decref(job);
+  pty_process_close_master(ptyproc);
+  Process *proc = (Process *)ptyproc;
+  if (proc->internal_close_cb) {
```
`proc->internal_close_cb` is not initialized at this point if the spawn fails.

tarruda replied:
:+1:

tarruda commented:
> I didn't see the OSX logs before you updated this PR, but we have been seeing random hangs on OSX on master. I just assumed it was a travis performance issue as it was in May.

The hangs in this branch are consistent, I just don't understand why they aren't present in master since the cause is seems to be the sleep program spawned in job_spec.lua

tarruda committed:
event loop: New abstraction layer with refactored time/signal API

- Add event loop abstraction module under src/nvim/event. The
  src/nvim/event/loop module replaces src/nvim/os/event
- Remove direct dependency on libuv signal/timer API and use the new abstraction
  instead.
- Replace all references to uv_default_loop() by &loop.uv, a new global variable
  that wraps libuv main event loop but allows the event loop functions to be
  reused in other contexts.

tarruda committed:
rstream/wstream: Unify structures and simplify API

- Simplify RStream/WStream API and make it more consistent with libuv.
- Move into the event loop layer(event subdirectory)
- Remove uv_helpers module.
- Simplify job/process internal modules/API.
- Unify RStream and WStream into a single structure. This is necessary because
  libuv streams can be readable and writable at the same time(and because the
  uv_helpers.c hack to associate multiple streams with libuv handle was removed)
- Make struct definition public, allowing more flexible/simple memory
  management by users of the module.
- Adapt channel/job modules to cope with the changes.

tarruda committed:
server: Extract most logic into the new socket abstraction

- Move event loop code into event/socket
- Reimplement server.c on top of the new SocketWatcher class
- Adapt msgpack_rpc/channel.c

tarruda committed:
job: Replace by a better process abstraction layer

- New libuv/pty process abstraction with simplified API and no globals.
- Remove nvim/os/job*. Jobs are now a concept that apply only to programs
  spawned by vimscript job* functions.
- Refactor shell.c/channel.c to use the new module, which brings a number of
  advantages:
  - Simplified API, less code
  - No slots in the user job table are used
  - Not possible to acidentally receive data from vimscript
- Implement job table in eval.c, which is now a hash table with unilimited job
  slots and unique job ids.

tarruda commented:
Merging now

The quickbuild seems to have failed because it couldn't publish the test results, so I'm going to disregard: http://neovim-qb.szakmeister.net/build/2089/step_status

Mingw link seems to be broken, so I will ignore it too.

tarruda commented:
Done, thanks for the help.

Tomorrow I start integrating the threading code. With this clearer event loop separation, the changes will be more contained.

oni-link commented on src/nvim/event/rstream.c:
```diff
@@ -0,0 +1,167 @@
+      // Read error or EOF, either way stop the stream and invoke the callback
+      // with eof == true
+      uv_read_stop(uvstream);
+      stream->read_cb(stream, stream->buffer, stream->data, true);
```
`stream->read_cb` needs a `NULL` check before making the call. The default value for the variable is `NULL`.

tarruda replied:
Is there any reason for allowing NULL `read_cb`? How about putting an assertion or `FUNC_NONNULL_ATTR_ALL` in `rstream_start`?

oni-link replied:
In `do_os_system()` the output rstreams could be started with callback `NULL`.

tarruda replied:
:+1:

oni-link commented on src/nvim/event/pty_process.c:
```diff
@@ -120,23 +67,18 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
+  proc->pid = pid;
   return true;
 
 error:
```
When this label is reached, the child ignores the signal `SIGTERM` and we have to wait for 5 seconds until we can send signal `SIGKILL`.
If that happens, nvim hangs with
```
Vim: Caught deadly signal 'SIGTERM'

Vim: Finished.
```
Can this error handling be simplified to 
```c
close(master);
kill(pid,SIGKILL);
return false;
```
?

tarruda replied:
I suppose it can. This cleanup code was created in @a2b45357 so I opted to not mess with it right now.

@elmart, is there any specific reason you added this timeout?

elmart replied:
Well, intention was just giving some time for spawned child to quit orderly, responding to SIGTERM, and only use SIGKILL if no response after some time (KILL_RETRIES * KILL_TIMEOUT).

oni-link commented on src/nvim/eval.c:
```diff
@@ -15226,19 +15223,15 @@ static void f_termopen(typval_T *argvars, typval_T *rettv)
+  data->proc.pty.height = curwin->w_height;
+  data->proc.pty.term_name = xstrdup("xterm-256color");
+  if (!common_job_start(data, rettv)) {
     return;
```
TerminalJobData and term_name are not freed:
Cleanup could be:
```c
    free(data->proc.pty.term_name);
    free_term_job_data(data);

```

tarruda replied:
:+1:

tarruda replied:
Added this and some other fixes to the beginning of #3029

jszakmeister commented:
Quickbuild checks the outcome of the tests on the publish step, so it's not a good idea to ignore it.  One thing I really dislike about busted is that it doesn't have a nice setup for the junit style testing, and it mixes in other stuff into the XML output--which corrupts the document.  In this case, there was an error, and there was this emitted into the beginning of the document:

```text
Warning: Screen changes have been received after the expected state was seen.
This is probably due to an indeterminism in the test. Try adding
`wait()` (or even a separate `screen:expect(...)`) at a point of possible
indeterminism, typically in between a `feed()` or `execute()` which is non-
synchronous, and a synchronous api call.

Note that sometimes a `wait` can trigger redraws and consequently generate more
indeterminism. If adding `wait` calls seems to increase the frequency of these
messages, try removing every `wait` call in the test.

If everything else fails, use Screen:redraw_debug to help investigate what is
  causing the problem.
      
stack traceback:
	./test/functional/ui/screen.lua:270: in function 'wait'
	./test/functional/ui/screen.lua:219: in function 'expect'
	test/functional/ui/screen_basic_spec.lua:492: in function <test/functional/ui/screen_basic_spec.lua:490>
```

The test case that failed was this one:


```text
		<testcase name='Screen resize has minimum width/height values' classname='test/functional/ui/screen_basic_spec.lua:490'>
			<error>./test/functional/ui/screen.lua:276: Row 1 didn&apos;t match.
Expected: &quot;-- INS^ERT --&quot;
Actual:   &quot;resize^      &quot;
stack traceback:
	./test/functional/ui/screen.lua:276: in function &apos;wait&apos;
	./test/functional/ui/screen.lua:219: in function &apos;expect&apos;
	test/functional/ui/screen_basic_spec.lua:492: in function &lt;test/functional/ui/screen_basic_spec.lua:490&gt;
</error>
		</testcase>
```

tarruda commented:
@jszakmeister I admit that sometimes I have trouble parsing quickbuild output. This particular error was already present though, it's a indeterminism I failed to fix in this PR.

oni-link commented on src/nvim/eval.c:
```diff
@@ -10928,14 +10930,15 @@ static void f_jobstop(typval_T *argvars, typval_T *rettv)
 
-  if (!is_user_job(job)) {
+  TerminalJobData *data = pmap_get(uint64_t)(jobs, argvars[0].vval.v_number);
+  if (!data || data->stopped) {
```
Most of the other checks for a valid "job" do not check for `data->stopped`. Should probably also be added there.

tarruda replied:
:+1: 

https://github.com/tarruda/neovim/commit/98a6925350551696fedf03328da376ebc8e37d09

jszakmeister commented:
> @jszakmeister I admit that sometimes I have trouble parsing quickbuild output.

Me too.  I wish it QuickBuild was better in that regard.

oni-link commented on src/nvim/main.c:
```diff
@@ -133,6 +137,41 @@ static const char *err_extra_cmd =
 
+void event_init(void)
+{
+  loop_init(&loop, NULL);
```
This call to `loop_init()` is too late. `command_line_scan()` in `main()` could already need the loop initialized.
See https://github.com/neovim/neovim/issues/3045#issuecomment-123405833.

tarruda replied:
:+1:

mhinz replied:
:+1:
